File: python/tests/helpers/stdio_child.py
from __future__ import annotations

import pathlib
import sys


def main() -> int:
    mode = sys.argv[1]

    if mode == "echo":
        for line in sys.stdin:
            sys.stdout.write(line)
            sys.stdout.flush()
        return 0

    if mode == "crash-after-first-line":
        line = sys.stdin.readline()
        if line:
            sys.stderr.write("child saw line before crash\n")
            sys.stderr.flush()
        return 7

    if mode == "emit-path-and-exit":
        sys.stdout.write(str(pathlib.Path.cwd()) + "\n")
        sys.stdout.flush()
        return 0

    raise ValueError(f"Unsupported mode: {mode}")


if __name__ == "__main__":
    raise SystemExit(main())

File: typescript/tests/helpers/stdio-child.mjs
const mode = process.argv[2];

if (mode === "echo") {
  process.stdin.setEncoding("utf8");
  process.stdin.on("data", (chunk) => {
    process.stdout.write(chunk);
  });
  process.stdin.on("end", () => {
    process.exit(0);
  });
} else if (mode === "crash-after-first-line") {
  process.stdin.setEncoding("utf8");
  let buffer = "";
  process.stdin.on("data", (chunk) => {
    buffer += chunk;
    if (buffer.includes("\n")) {
      process.stderr.write("child saw line before crash\n");
      process.exit(7);
    }
  });
} else if (mode === "emit-path-and-exit") {
  process.stdout.write(`${process.cwd()}\n`);
  process.exit(0);
} else {
  throw new Error(`Unsupported mode: ${mode}`);
}
